pkg/ccl/backupccl/backupinfo/manifest_handling.go

```diff
@@ -943,21 +984,51 @@ func LoadSQLDescsFromBackupsAtTime(
 		}
 		lastBackupManifest = b
 	}
-	if len(lastBackupManifest.DescriptorChanges) == 0 {
-		descs, err := BackupManifestDescriptors(&lastBackupManifest)
-		return descs, lastBackupManifest, err
+
+	// From this point on we try to load descriptors based on descriptor
+	// revisions. The algorithm below assumes that descriptor revisions are sorted
+	// by descChangesLess, which is a sort by descriptor ID, then descending by
+	// revision time for revisions with the same ID. The external SST for
+	// descriptors already have entries sorted in this order, we just have to make
+	// sure the in-memory descriptors in the manifest are ordered as well.
+	sort.Slice(lastBackupManifest.DescriptorChanges, func(i, j int) bool {
+		return descChangesLess(&lastBackupManifest.DescriptorChanges[i], &lastBackupManifest.DescriptorChanges[j])
+	})
+
+	descRevIt := lastIterFactory.NewDescriptorChangesIter(ctx)
+	defer descRevIt.Close()
+	if ok, err := descRevIt.Valid(); err != nil {
+		return nil, backuppb.BackupManifest{}, err
+	} else if !ok {
+		descs, err := BackupManifestDescriptors(ctx, lastIterFactory, lastBackupManifest.EndTime)
+		if err != nil {
+			return nil, backuppb.BackupManifest{}, err
+		}
+		return descs, lastBackupManifest, nil
 	}
 
-	byID := make(map[descpb.ID]catalog.DescriptorBuilder, len(lastBackupManifest.Descriptors))
-	for _, rev := range lastBackupManifest.DescriptorChanges {
-		if asOf.Less(rev.Time) {
+	byID := make(map[descpb.ID]catalog.DescriptorBuilder, 0)
+	prevRevID := descpb.InvalidID
+	for ; ; descRevIt.Next() {
+		if ok, err := descRevIt.Valid(); err != nil {
+			return nil, backuppb.BackupManifest{}, err
+		} else if !ok {
 			break
 		}
-		if rev.Desc == nil {
-			delete(byID, rev.ID)
-		} else {
+
+		rev := descRevIt.Value()
+		if asOf.Less(rev.Time) {
+			continue
+		}
+
+		if rev.ID == prevRevID {
```

Comment:
maybe it's just me but the sorting comment is a little far from this logic. Can I bug you to add a sentence that since revisions are sorted by ID and the descending order of revisions we can skip over older revisions once we have seen a revision for an ID.